contracts/NonFungibleToken.cdc

```diff
@@ -41,162 +36,195 @@ Collection to complete the transfer.
+    /// because only code within a collection or that has special entitled access
+    /// to the collections methods will be able to get the entitled reference
+    /// 
+    /// The event makes it so that third-party indexers can monitor the events
+    /// and query the updated metadata from the owners' collections.
     ///
-    pub event ContractInitialized()
+    access(all) event Updated(id: UInt64, uuid: UInt64, owner: Address?, type: String)
+    access(all) view fun emitNFTUpdated(_ nftRef: auth(Update | Owner) &{NonFungibleToken.NFT})
+    {
+        emit Updated(id: nftRef.id, uuid: nftRef.uuid, owner: nftRef.owner?.address, type: nftRef.getType().identifier)
+    }
+
 
     /// Event that is emitted when a token is withdrawn,
-    /// indicating the owner of the collection that it was withdrawn from.
+    /// indicating the type, id, uuid, the owner of the collection that it was withdrawn from,
+    /// and the UUID of the resource it was withdrawn from, usually a collection.
     ///
     /// If the collection is not in an account's storage, `from` will be `nil`.
     ///
-    pub event Withdraw(id: UInt64, from: Address?)
+    access(all) event Withdrawn(type: String, id: UInt64, uuid: UInt64, from: Address?, providerUUID: UInt64)
 
     /// Event that emitted when a token is deposited to a collection.
+    /// Indicates the type, id, uuid, the owner of the collection that it was deposited to,
+    /// and the UUID of the collection it was deposited to
     ///
-    /// It indicates the owner of the collection that it was deposited to.
+    /// If the collection is not in an account's storage, `from`, will be `nil`.
     ///
-    pub event Deposit(id: UInt64, to: Address?)
+    access(all) event Deposited(type: String, id: UInt64, uuid: UInt64, to: Address?, collectionUUID: UInt64)
 
-    /// Interface that the NFTs have to conform to
-    /// The metadata views methods are included here temporarily
-    /// because enforcing the metadata interfaces in the standard
-    /// would break many contracts in an upgrade. Those breaking changes
-    /// are being saved for the stable cadence milestone
+    /// Included for backwards-compatibility
+    access(all) resource interface INFT: NFT {}
+
+    /// Interface that the NFTs must conform to
     ///
-    pub resource interface INFT {
-        /// The unique ID that each NFT has
-        pub let id: UInt64
+    access(all) resource interface NFT: ViewResolver.Resolver {
 
-        /// Function that returns all the Metadata Views implemented by a Non Fungible Token
-        ///
-        /// @return An array of Types defining the implemented views. This value will be used by
-        ///         developers to know which parameter to pass to the resolveView() method.
-        ///
-        pub fun getViews(): [Type] {
-            return []
+        /// unique ID for the NFT
+        access(all) let id: UInt64
```

Comment:
I feel like I know the answer here, but I have to ask - with the introduction of Flow EVM, the ERC721 standard has me wondering if there's any desire to update `id` from `UInt64` to `UInt256`? Selfishly asking as it will otherwise be a potential pain point for bridged assets, not that changing it at this point wouldn't have it's difficulty & effort as it would likely require a migration.

Reply:
I have no `attachment` to UInt64 for the id type, so if it is possible in a migration, then I'm okay with it. @turbolent What do you think? would it be possible to migrate all ID fields of NFTs to `UInt256`? We would potentially need to migrate uuids also, but I'm not sure.

@sisyphusSmiling It would also be good to understand how much of a headache it is for you to have this mismatch. If it is something you can handle relatively easily, then it might not be worth the hassle for trying to figure out a way to migrate safely if that is going to be difficult.

Reply:
I'm trying to avoid pigeonholing IDs for EVM-native NFTs bridging to Flow. The problem fundamentally comes down to the fact that ERC721 ids are `uint256`, meaning we have 1/4 available ids in Flow than might exist in EVM.

This presents two problems:
1. Accurately representing the NFT ID on Flow according to the NFT standard without collisions
2. Ensuring IDs that bridge to Flow are bridged back to their correct corresponding NFT

The only way I can think of to handle this is by having a `mod` field on bridged EVM-native NFTs. So imagine an ERC721 with id = UInt64.max + 1 would look like:

```cadence
access(all) resource NFT {
    access(all) let id: UInt64 = 1
    access(all) let mod: UInt8 = 1
}
```

where id = ERC721.id % UInt64.max and mod = ERC721.id / UInt64.max. I'm effectively wrapping the UInt64 based on the UInt256 value.

This solves 2 as I can recalculate the original `uint256` value from these two fields. However, I'm still left with the problem of 1 - collisions on IDs once NFTs are bridged to Flow are unavoidable.

Reply:
What backend systems even support UInt256? A change like that could cause havoc in lots of places, does it makes sense to move this to a FLIP instead of here in a PR, or to schedule a discussion? It would have massive implications for every existing app that's already having to do a lot of work for 1.0

Reply:
Given the priority on finalizing the standards, I don't want to hold up progress. I raised it here as a point of discussion and to temp check sentiment on such a change, so thanks for adding a point of concern @austinkline. I don't doubt there would be impacts to codebases outside of the context of just this standard, but figured this is likely the only chance we'll have to make significant changes and this hasn't yet been discussed.

@joshuahannan based on what we have to discuss on Thursday, would you recommend raising this in our sync then? Meanwhile, happy to discuss async as I know the timeline for v2 standard delivery is compressed and high priority.

Reply:
@sisyphusSmiling I think part of this will depend on how you envision an NFT being represented across chains here, because you could always let an EVM-origin nft use its resource uuid as the ID, and then maintain an EVM id field that has the proper number

Reply:
Great point @austinkline! Adding a field mapping to the EVM-defined ID makes sense in a world with different id types. And you're right, much less work for all involved to create new guidance moving forward than refactor existing code and offchain systems